pep-0635.rst

```diff
@@ -0,0 +1,689 @@
+in turn, gives rise to a high degree of versatility by favouring a
+'duck typing' style, imposing only the bare minimum of requirements
+on the form and shape of data, i.e. the structure of data objects.
+Nonetheless, we find that Python code is often sparkled with
+conditions depending on ``isinstance``, ``len``, ``getattr``,
+``hasattr``, etc.  Despite the benefits of 'duck typing', actual
+code still requires to query the format or type of an object in order
+to choose an appropriate processing action or to extract relevant bits
+of information.
+
+Unfortunately, the conditions for slightly more complex structures
+quickly grow into sequences of partly interdependent bits of
+structural tests, hurting readbility and maintability.  Pattern
+matching offers here a more direct scheme of expressing the minimal
+structure that an objects needs to have in order to allow for further
+processing.  Rather than writing a series of manual tests, patterns
+follow a *declarative* style, which is well-known for improved
+readability, maintability, and for delegating the burden of efficient
+execution on the compiler and interpreter.
+
+The concept of pattern matching is similar to regular expressions,
+where succinct patterns describe a textual structure.  A dedicated
+compiler then transforms these declarative patterns into highly
+efficient finite state machines.  In contrast to regular expressions,
+pattern matching targets Python objects rather than textual data,
+and builds on *decision trees* as the driving motor for finding a
+match.  Moreover, pattern matching blends the matching of a suitable
+pattern with that of a function, i.e. code that is executed in order
+to handle and process the information provided by a specific kind of
+object.
+
+One of the simplest forms of pattern matching that we find in other
+languages comes in the form of *function overloading*.  The type and
+number of arguments then determine which implementation of a specific
+function will be executed.  Object-oriented languages (including
+Python) may also use the *visitor pattern* to differentiate an action
+based on the type or class of an object.  Both of these approaches,
+however, are aimed at 'shallow' structures with little or no direct
+support for nested structures or structural information that is not
+directly encoded in an object's class or type.  For instance, it is
+simple to differentiate between an integer, a string, and a tuple, say,
+but becomes quite cumbersome and difficult to differentiate between
+tuples of different lengths, or between one containing string elements
+vs. one containing numeric elements.  This is where pattern matching
+shines: for structures that go beyond simple class distinctions.
+
+Although pattern matching is a concept that has been known and used
+for decades, we propose a re-interpretation that centres around the
+principle of 'duck typing' and builds on existing features of the
```

Comment:
How exactly do we leverage duck typing? Most of our checks actually rely on isinstance().

Reply:
With the attribute-checks (e.g. `object(spam=P, egg=Q)`) we accept any object that has the specified attributes (which in turn match the respective patterns).  I feel this is an important addition for a dynamic language like Python.  The custom match protocol would have been the second important step towards that, although we deferred that.

Reply:
(As I said in another comment that's an edge case of the class pattern and IMO shouldn't be the core of our argument.)